PayloadPkg/FirmwareUpdate/FirmwareUpdateHelper.c

```diff
@@ -18,7 +18,193 @@
+
+**/
+EFI_STATUS
+EFIAPI
+BootMediaWrite (
+  IN     UINT64   Address,
+  IN     UINT32   ByteCount,
+  OUT    UINT8    *Buffer
+  )
+{
+  return mFwuSpiService->SpiWrite (FlashRegionBios, (UINT32)Address, ByteCount, Buffer);
+}
+
+/**
+  This function erases blocks from the SPI device.
+
+  @param[in]  Address             The block address in the FlashRegionAll to read from on the SPI.
+  @param[in]  ByteCount           Size of the region to erase in bytes.
+
+  @retval EFI_SUCCESS             Erase completes successfully.
+  @retval others                  Device error, the command aborts abnormally.
+
+**/
+EFI_STATUS
+EFIAPI
+BootMediaErase (
+  IN     UINT64   Address,
+  IN     UINT32   ByteCount
+  )
+{
+  return mFwuSpiService->SpiErase (FlashRegionBios, (UINT32)Address, ByteCount);
+}
+
+/**
+  Get SVN from existing firmware
+
+  This routine get SPI base address and read first four bytes
+  to get STAGE1A FV base address, using this base address
+  this routine can calculate the offset to the SVN structure.
+
+  @param[in]  Stage1AFvPointer    Pointer to Stage1A fv base.
+  @param[out] BlVersion           Pointer to SBL version struct.
+
+  @retval EFI_SUCCESS             Read SVN success
+  @retval EFI_INVALID_PARAMETER   Invalid parameter
+**/
+EFI_STATUS
+EFIAPI
+GetSvn (
+  IN  UINT32                Stage1AFvPointer,
+  OUT BOOT_LOADER_VERSION   **BlVersion
+  )
+{
+  UINT32                Stage1AFvBase;
+  UINT32                TopSwapRegionSize;
+  EFI_STATUS            Status;
+
+  //
+  // When updating BP1 - read version from backup partition
+  // When updatint BP0 - since top swap bit is set, we need
+  //                     to substract the top swap region size
+  //                     because when top swap bit is set
+  //                     address lines will be inverted and point
+  //                     to stage 1A from primary partition.
+  //
+  GetRegionInfo(&TopSwapRegionSize, NULL, NULL);
+  Stage1AFvPointer = Stage1AFvPointer - TopSwapRegionSize;
+
+  Stage1AFvBase = (UINT32)(*(UINT32 *)(UINTN)Stage1AFvPointer);
+  Stage1AFvBase = Stage1AFvPointer - (~Stage1AFvBase + 1) + sizeof(UINT32);
+
+  Status = GetVersionfromFv (&Stage1AFvBase, FALSE, BlVersion);
+  if (EFI_ERROR (Status)) {
+    DEBUG((DEBUG_ERROR, "Getting firmware version failed with status: %r\n", Status));
+    return Status;
+  }
+
+  return Status;
+}
+
+/**
+  Verify the firmware version to make sure it is no less than current firmware version.
+
+  @param[in]  Stage1ABase   Pointer to stage 1A base.
+  @param[in]  IsFd          Does Stage1ABase point to Stage1A FD
+                            or SBL Stage1A FV ?
+  @param[out] Version       Pointer to version of the firmware
+
+  @retval  EFI_SUCCESS        The operation completed successfully.
+  @retval  others             There is error happening.
+**/
+EFI_STATUS
+GetVersionfromFv (
+  IN  UINT32              *Stage1ABase,

```

Comment:
Why do you need pointer here ?  Since we don't change the value of the Stage1ABase, we should just pass the value directly.

Reply:
changed code to just pass the base